deploy/s3.sh

```diff
@@ -0,0 +1,189 @@
+  # fullchain
+  _deploy_to_bucket $_cfullchain "$_cdomain/fullchain.cer"
+
+  return 0
+
+}
+
+####################  Private functions below ##################################
+
+_deploy_to_bucket() {
+  if [ "$_aws_cli_installed" -eq "0" ]; then
+    _deploy_with_curl $1 $2
+  else
+    _deploy_with_awscli $1 $2
+  fi
+}
+
+_deploy_with_awscli() {
+  file="$1"
+  bucket="$S3_BUCKET"
+  prefix="$2"
+  region="$S3_REGION"
+
+  aws s3 cp "$file" s3://"$bucket"/"$prefix" --region "$region"
+}
+
+_deploy_with_curl() {
+
+  file="${1}"
+  bucket="${S3_BUCKET}"
+  prefix="${2}"
+  region="${S3_REGION}"
+  acl="private"
+  timestamp="$(date -u "+%Y-%m-%d %H:%M:%S")"
+  signed_headers="date;host;x-amz-acl;x-amz-content-sha256;x-amz-date"
+
+  if [[ $(uname) == "Darwin" ]]; then
+    iso_timestamp=$(date -ujf "%Y-%m-%d %H:%M:%S" "${timestamp}" "+%Y%m%dT%H%M%SZ")
+    date_scope=$(date -ujf "%Y-%m-%d %H:%M:%S" "${timestamp}" "+%Y%m%d")
+    date_header=$(date -ujf "%Y-%m-%d %H:%M:%S" "${timestamp}" "+%a, %d %h %Y %T %Z")
+  else
+    iso_timestamp=$(date -ud "${timestamp}" "+%Y%m%dT%H%M%SZ")
+    date_scope=$(date -ud "${timestamp}" "+%Y%m%d")
+    date_header=$(date -ud "${timestamp}" "+%a, %d %h %Y %T %Z")
+  fi
+
+  _info "Uploading $S3_BUCKET/$prefix"
+
+  curl \
+    -T "${file}" \
+    -H "Authorization: AWS4-HMAC-SHA256 Credential=${AWS_ACCESS_KEY_ID}/${date_scope}/${region}/s3/aws4_request,SignedHeaders=${signed_headers},Signature=$(_signature)" \
+    -H "Date:${date_header}" \
+    -H "x-amz-acl:${acl}" \
+    -H "x-amz-content-sha256:$(_payload_hash)" \
+    -H "x-amz-date:${iso_timestamp}" \
+    "https://${bucket}.s3.${region}.amazonaws.com/${prefix}"
+
+}
+
+_payload_hash() {
+  local output=$(shasum -ba 256 "$file")
+  echo "${output%% *}"
+}
+
+_canonical_request() {
+  echo "PUT"
+  echo "/${prefix}"
+  echo ""
+  echo "date:${date_header}"
+  echo "host:${bucket}.s3.${region}.amazonaws.com"
+  echo "x-amz-acl:${acl}"
+  echo "x-amz-content-sha256:$(_payload_hash)"
+  echo "x-amz-date:${iso_timestamp}"
+  echo ""
+  echo "${signed_headers}"
+  printf "$(_payload_hash)"
+}
+
+_canonical_request_hash() {
+  local output=$(_canonical_request | shasum -a 256)
+  echo "${output%% *}"
+}
+
+_string_to_sign() {
+  echo "AWS4-HMAC-SHA256"
+  echo "${iso_timestamp}"
+  echo "${date_scope}/${region}/s3/aws4_request"
+  printf "$(_canonical_request_hash)"
+}
+
+_signature_key() {
+  local secret=$(printf "AWS4${AWS_SECRET_ACCESS_KEY?}" | _hex_key)
+  local date_key=$(printf ${date_scope} | _hmac_sha256 "${secret}" | _hex_key)
+  local region_key=$(printf ${region} | _hmac_sha256 "${date_key}" | _hex_key)
+  local service_key=$(printf "s3" | _hmac_sha256 "${region_key}" | _hex_key)
+  printf "aws4_request" | _hmac_sha256 "${service_key}" | _hex_key
+}
+
+_hex_key() {
+  hexdump -ve '1/1 "%.2x"'; echo
+}
+
+_hmac_sha256() {
+  local hexkey=$1
+  openssl dgst -binary -sha256 -mac HMAC -macopt hexkey:${hexkey}
```

Comment:
do not use `openssl` directly,  we have a function `_hmac()`, please use that instead.